### DISCUSSION TITLE
How to ensure every player is online before creating groups.

### DISCUSSION THREAD

**Discussion by jongliko** [hklages/node-red-contrib-sonos-plus#301] [Q&A] [answered]
Hi,

Is there a way to test if players are online before grouping ? Or is there a way to avoid errors if one player is missing (offline) when using household.create.group and just group the available (online) players.

I tried using household.test.player before creating the group but it seems not to be the ideal way.

Thanks for your help

**Comment by hklages** [comment, marked as answer]
Hi. 

Best is to keep the SONOS player always on   - that's why SONOS players do not have a build in on/off switch.  :-)  (except the Roam)

If you switch on or/off a SONOS player via power supply it takes several minutes until the new status is updated. 

Option 0: Don't swich off - easiest to implement and 100% reliable

Option 1: You wait 15 minutes (or maybe also 10 minutes or even shorter). After that period the standard commands such as household.test.player or household.get.groups should provide the correct data (no garantee - it is in the hand of your network and SONOS).

Option 2: Install the sister package sonos-events and listen to the group event. Store the data in a flow or global variable. 
On any change compare the new group data with the previous and check whether you player is now available and then start add the player to the group. Here is an example flow.  That usually works for me.

Does that answer help? 


```
[
    {
        "id": "79bd7e5a.995ee",
        "type": "change",
        "z": "4b75157e.b475ec",
        "name": "current /previous player name",
        "rules": [
            {
                "t": "set",
                "p": "previousPlayerArray",
                "pt": "msg",
                "to": "$flowContext('previousPlayerArray')",
                "tot": "jsonata"
            },
            {
                "t": "set",
                "p": "currentPlayerArray",
                "pt": "msg",
                "to": "payload.**.playerName",
                "tot": "jsonata"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 401,
        "y": 990,
        "wires": [
            [
                "27455f3c.cab9b"
            ]
        ]
    },
    {
        "id": "2788e067.82b4b",
        "type": "inject",
        "z": "4b75157e.b475ec",
        "name": "init",
        "props": [
            {
                "p": "payload"
            },
            {
                "p": "topic",
                "vt": "str"
            }
        ],
        "repeat": "",
        "crontab": "",
        "once": false,
        "onceDelay": 0.1,
        "topic": "",
        "payload": "",
        "payloadType": "date",
        "x": 472,
        "y": 957,
        "wires": [
            [
                "41f87640.5995a8"
            ]
        ]
    },
    {
        "id": "41f87640.5995a8",
        "type": "change",
        "z": "4b75157e.b475ec",
        "name": "",
        "rules": [
            {
                "t": "set",
                "p": "previousPlayerArray",
                "pt": "flow",
                "to": "[\"SonosKueche\"]",
                "tot": "json"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 660,
        "y": 957,
        "wires": [
            []
        ]
    },
    {
        "id": "72b3ae2a.1363",
        "type": "change",
        "z": "4b75157e.b475ec",
        "name": "update flow variable",
        "rules": [
            {
                "t": "set",
                "p": "previousPlayerArray",
                "pt": "flow",
                "to": "currentPlayerArray",
                "tot": "msg"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 790,
        "y": 1040,
        "wires": [
            []
        ]
    },
    {
        "id": "c80e53f6.c3166",
        "type": "switch",
        "z": "4b75157e.b475ec",
        "name": "SonosWintergarten new?",
        "property": "payload",
        "propertyType": "msg",
        "rules": [
            {
                "t": "jsonata_exp",
                "v": "$count($filter(payload.[$='SonosWintergarten'], function ($item) {\t    $item\t})) > 0",
                "vt": "jsonata"
            }
        ],
        "checkall": "false",
        "repair": false,
        "outputs": 1,
        "x": 716,
        "y": 990,
        "wires": [
            [
                "d99a6d.4f47159"
            ]
        ]
    },
    {
        "id": "27455f3c.cab9b",
        "type": "function",
        "z": "4b75157e.b475ec",
        "name": "get new player list",
        "func": "// msg.currentPlayerArray current player array\n// msg.previousPlayerArray previous player array\n\n// returns array with names of those not in previous but in current\nconst current = msg.currentPlayerArray\nconst previous = msg.previousPlayerArray\n\nconst newPlayers = []\ncurrent.forEach(player => {\n    if (!previous.includes(player)) {\n        newPlayers.push(player)   \n    }\n})\nmsg.payload = newPlayers\nreturn msg",
        "outputs": 1,
        "noerr": 0,
        "initialize": "",
        "finalize": "",
        "libs": [],
        "x": 550,
        "y": 1040,
        "wires": [
            [
                "72b3ae2a.1363",
                "c80e53f6.c3166",
                "f997b332a8794e57"
            ]
        ]
    },
    {
        "id": "b3228ed3.b10fb",
        "type": "sonos-universal",
        "z": "4b75157e.b475ec",
        "confignode": "88e9fa70.4a53e8",
        "command": "household.create.group",
        "state": "SonosKueche,SonosBad,SonosWohnzimmer,SonosWintergarten",
        "stateType": "str",
        "avoidCheckPlayerAvailability": false,
        "name": "Add SonosWintergarten",
        "x": 1114,
        "y": 990,
        "wires": [
            []
        ]
    },
    {
        "id": "d99a6d.4f47159",
        "type": "delay",
        "z": "4b75157e.b475ec",
        "name": "",
        "pauseType": "delay",
        "timeout": "5",
        "timeoutUnits": "seconds",
        "rate": "1",
        "nbRateUnits": "1",
        "rateUnits": "second",
        "randomFirst": "1",
        "randomLast": "5",
        "randomUnits": "seconds",
        "drop": false,
        "outputs": 1,
        "x": 915,
        "y": 990,
        "wires": [
            [
                "b3228ed3.b10fb"
            ]
        ]
    },
    {
        "id": "7fc21158.550ed",
        "type": "sonosevents-selection",
        "z": "4b75157e.b475ec",
        "name": "",
        "confignode": "51969b5e.59dda4",
        "playerHostname": "192.168.178.51",
        "events": [
            {
                "fullName": "ZoneGroupTopologyService.allGroups"
            }
        ],
        "outputs": 1,
        "x": 145,
        "y": 990,
        "wires": [
            [
                "79bd7e5a.995ee"
            ]
        ]
    },
    {
        "id": "83e23360.b9a37",
        "type": "comment",
        "z": "4b75157e.b475ec",
        "name": "Add Wintergarden to group when getting online",
        "info": "",
        "x": 185,
        "y": 957,
        "wires": [],
        "icon": "node-red/alert.svg"
    },
    {
        "id": "f997b332a8794e57",
        "type": "switch",
        "z": "4b75157e.b475ec",
        "name": "SonosWohnzimmer new?",
        "property": "payload",
        "propertyType": "msg",
        "rules": [
            {
                "t": "jsonata_exp",
                "v": "$count($filter(payload.[$='SonosWohnzimmer'], function ($item) {\t    $item\t})) > 0",
                "vt": "jsonata"
            }
        ],
        "checkall": "false",
        "repair": false,
        "outputs": 1,
        "x": 716,
        "y": 1079,
        "wires": [
            [
                "9f491749af71872d"
            ]
        ]
    },
    {
        "id": "9f491749af71872d",
        "type": "delay",
        "z": "4b75157e.b475ec",
        "name": "",
        "pauseType": "delay",
        "timeout": "5",
        "timeoutUnits": "seconds",
        "rate": "1",
        "nbRateUnits": "1",
        "rateUnits": "second",
        "randomFirst": "1",
        "randomLast": "5",
        "randomUnits": "seconds",
        "drop": false,
        "outputs": 1,
        "x": 915,
        "y": 1079,
        "wires": [
            [
                "4486c43e57e765bb"
            ]
        ]
    },
    {
        "id": "4486c43e57e765bb",
        "type": "sonos-universal",
        "z": "4b75157e.b475ec",
        "confignode": "88e9fa70.4a53e8",
        "command": "household.create.group",
        "state": "SonosKueche,SonosBad,SonosWohnzimmer",
        "stateType": "str",
        "avoidCheckPlayerAvailability": false,
        "name": "Add SonosWohnzimmer",
        "x": 1114,
        "y": 1079,
        "wires": [
            []
        ]
    },
    {
        "id": "88e9fa70.4a53e8",
        "type": "sonos-config",
        "name": "player_kitchen_dns",
        "serialnum": "",
        "ipaddress": "sonoskitchen.fritz.box"
    },
    {
        "id": "51969b5e.59dda4",
        "type": "sonosevents-config",
        "name": "household",
        "listenerHostname": "",
        "listenerPort": "",
        "portType": "num"
    }
]
```

**Comment by jongliko** [comment]
Hi,

Thanks for your answer and for this amazing work. 

I will try tonight the solution 2 you gave me.

I have problems with Sonos Move that goes offline when  they are not playing and not charging.

And sometimes a player is not responding or has been unplugged by someone in the house.

I have seen that the sister package sonos-events will not be actively developed in the future. Does it mean that if something break in the future, it will not be fixed ? 

What do yout think about household.test.player accepting more than one player . IN string like "Player1,Player2,Player3" and Out True/False.

Thanks again

**Comment by hklages** [comment]
> I have seen that the sister package sonos-events will not be actively developed in the future. Does it mean that if something break in the future, it will not be fixed ?

That means, that I most likely won't add new features and leave it as it is. My intention is to maintain it for some time and fix bugs and update dependencies.  I am 65 and already retired - so dont know how long I will have the passion, skills, health to maintain it. 

> What do yout think about household.test.player accepting more than one player . IN string like "Player1,Player2,Player3" and Out True/False.

Node-RED can handle that. Just take household.get.groups. The result is an array of groups where the groups are an array of  players. 
You can use JSONATA to flatten that to a simple array and then start your search Player1, .. in that array. 
Here is a similiar [solution](https://stackoverflow.com/questions/73882914/how-to-flatten-array-of-arrays-in-json-using-jsonata).

If you dont like JSONATA you can write a Javascript function. 

test.player is a "work around" - the result depends very much on the number of players in the network and the network performance. 
I don't want to spend to much time on it.

**Reply by jongliko** [reply]
Thanks for your answers. It's perfect ! I think I now have plenty of solutions to work with. 

Thanks again for all the time you spend on this amazing contribs, and I'm sure that passion, skills and health will always follow you ;)

**Comment by jongliko** [comment]
Ok here what I have done with your help. Using household.get.groups and JSONATA to flatten an array containing only "playerName".

![Capture d'écran 2024-07-04 110402](https://github.com/hklages/node-red-contrib-sonos-plus/assets/5091679/7b52c60d-d2de-4b94-8631-0d3478371f8e)

I added a little function to check if players are in the array. Maybe it could be done only with JSONATA.

`[
    {
        "id": "1e2139b8227fac80",
        "type": "inject",
        "z": "cf4287a2c5ae5b86",
        "name": "Trigg",
        "props": [
            {
                "p": "payload"
            },
            {
                "p": "topic",
                "vt": "str"
            }
        ],
        "repeat": "",
        "crontab": "",
        "once": false,
        "onceDelay": 0.1,
        "topic": "",
        "payload": "",
        "payloadType": "date",
        "x": 1490,
        "y": 1440,
        "wires": [
            [
                "a3eb280ff2286edd"
            ]
        ]
    },
    {
        "id": "a3eb280ff2286edd",
        "type": "sonos-universal",
        "z": "cf4287a2c5ae5b86",
        "confignode": "eb10c4daf8df917a",
        "command": "household.get.groups",
        "state": "",
        "stateType": "str",
        "avoidCheckPlayerAvailability": false,
        "name": "",
        "x": 1660,
        "y": 1440,
        "wires": [
            [
                "b3ad99633a745ab1"
            ]
        ]
    },
    {
        "id": "af57f1bc49efbdfc",
        "type": "debug",
        "z": "cf4287a2c5ae5b86",
        "name": "Boolean",
        "active": true,
        "tosidebar": true,
        "console": false,
        "tostatus": false,
        "complete": "payload",
        "targetType": "msg",
        "statusVal": "",
        "statusType": "auto",
        "x": 2260,
        "y": 1440,
        "wires": []
    },
    {
        "id": "b3ad99633a745ab1",
        "type": "change",
        "z": "cf4287a2c5ae5b86",
        "name": "playerName Array",
        "rules": [
            {
                "t": "set",
                "p": "payload",
                "pt": "msg",
                "to": "*.playerName",
                "tot": "jsonata"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 1870,
        "y": 1440,
        "wires": [
            [
                "ab99329b86b78d11"
            ]
        ]
    },
    {
        "id": "ab99329b86b78d11",
        "type": "function",
        "z": "cf4287a2c5ae5b86",
        "name": "are Players In Array",
        "func": "const includesAll = (arr, values) => values.every(v => arr.includes(v));\nlet result = includesAll(msg.payload, ['player_1', 'player_2']); \nmsg = { payload: result};\nreturn msg;",
        "outputs": 1,
        "timeout": 0,
        "noerr": 0,
        "initialize": "",
        "finalize": "",
        "libs": [],
        "x": 2080,
        "y": 1440,
        "wires": [
            [
                "af57f1bc49efbdfc"
            ]
        ]
    },
    {
        "id": "c5fd7a6612f73781",
        "type": "comment",
        "z": "cf4287a2c5ae5b86",
        "name": "are All Players Alive",
        "info": "",
        "x": 1510,
        "y": 1400,
        "wires": []
    },
    {
        "id": "eb10c4daf8df917a",
        "type": "sonos-config",
        "name": "Cuisine",
        "serialnum": "",
        "ipaddress": "192.168.7.89"
    }
]`

**Comment by hklages** [comment]
Great  - usually a combination of JSONATA and then a function is the fastest way to find a solution. JSONATA is very powerful but sometime difficult to understand. 

If the payload contains the flattenend group output, then 
`$count($distinct($append(msg.payload, msg.onoffplayer))) = $count(msg.payload) ` as JSONATA might also work - but havent tested it
